scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -62,7 +60,41 @@ import scala.concurrent.{Future, Promise}
 import scala.io.Source
 import scala.reflect.ClassTag
 import scala.reflect.runtime.universe._
-import scala.util.Try
+import scala.util.{Failure, Success, Try}
+import scala.util.control.NonFatal
+
+/** Runner specific context. */
+trait RunnerContext {
+  def prepareOptions(options: PipelineOptions, artifacts: List[String]): Unit
+}
+
+private case object NoOpContext extends RunnerContext {
+  override def prepareOptions(options: PipelineOptions, artifacts: List[String]): Unit = Unit
+}
+
+/** Direct runner specific context. */
+private case object DirectContext extends RunnerContext {
+  override def prepareOptions(options: PipelineOptions, artifacts: List[String]): Unit = Unit
+}
+
+/** Companion object for [[RunnerContext]]. */
+private object RunnerContext {
+  private val mapping = Map(
+    "DirectRunner" -> DirectContext.getClass.getName,
+    "DataflowRunner" -> "com.spotify.scio.runners.dataflow.DataflowContext$")
+    .withDefaultValue(NoOpContext.getClass.getName)
+
+  // FIXME: this is ugly, is there a better way?
+  private def get(options: PipelineOptions): RunnerContext =
+    Class
+      .forName(mapping(options.getRunner.getName))
```

Comment:
Have some proper `try` and error massage in case of missing class? For example that maybe some runner dependency is missing etc.

Reply:
It's getting one of the value strings in `mapping` with a fallback `NoOpContext` so this should never fail.

Reply:
Unless classnames change - like `"com.spotify.scio.runners.dataflow.DataflowContext$"`, or that class is for some reason now available.

Reply:
Actually `DataflowResult` can be re-constructed with just project and job ID, async, and independent of submission so maybe keep it simple here and let caller (some centralized service that monitors job completion) handle retry/fallback?